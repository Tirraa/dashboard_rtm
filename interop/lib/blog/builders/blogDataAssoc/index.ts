import type { DocumentsTypesMetadatas } from '##/types/hell/contentlayerConfig';
import getBlogDataVariableName from './getBlogDataVariableName';

export function blogDataAssocBuilder(documentsTypesMetadatas: DocumentsTypesMetadatas) {
  const blogDataAssoc = {} as Record<string, string>;

  Object.keys(documentsTypesMetadatas).forEach((name) => {
    const currentMetadatas = documentsTypesMetadatas[name as keyof DocumentsTypesMetadatas];
    blogDataAssoc[currentMetadatas.categoryFolder] = getBlogDataVariableName(name);
  });

  return blogDataAssoc;
}

export default blogDataAssocBuilder;
