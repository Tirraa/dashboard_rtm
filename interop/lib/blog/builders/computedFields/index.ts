import buildBlogPostCategory from './functions/category';
import buildBlogPostLanguageFlag from './functions/language';
import buildBlogPostSlug from './functions/slug';
import buildBlogPostSubcategory from './functions/subcategory';
import buildBlogPostUrl from './functions/url';

export { buildBlogPostCategory, buildBlogPostLanguageFlag, buildBlogPostSlug, buildBlogPostSubcategory, buildBlogPostUrl };
