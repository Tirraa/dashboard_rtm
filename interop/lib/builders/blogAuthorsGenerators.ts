import type { EmptyString } from '@rtm/shared-types/CustomUtilityTypes';
import type { Index, Id } from '@rtm/shared-types/Numbers';

const emptyString: EmptyString = '';

export const generateIndexedAuthorNames = <T extends readonly string[]>(authorNames: T): Record<T[Index], Id> =>
  authorNames.reduce(
    (acc, tag, index) => {
      acc[tag] = index;
      return acc;
    },
    {} as Record<string, Id>
  ) as Record<T[Index], Id>;

export const generateBlogAuthorOptionsVocabSchema = <T extends readonly string[]>(authorNames: T): Record<T[Index], AuthorVocabArborescence> =>
  authorNames.reduce(
    (acc, tag) => {
      acc[tag] = { bio: emptyString };
      return acc;
    },
    {} as Record<string, AuthorVocabArborescence>
  ) as Record<T[Index], AuthorVocabArborescence>;

type AuthorVocabArborescence = { bio: EmptyString };
