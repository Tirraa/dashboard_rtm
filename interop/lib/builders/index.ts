import buildLandingPageLanguageFlag from './landing-pages/computedFields/functions/language';
import buildLandingPageCategory from './landing-pages/computedFields/functions/category';
import buildBlogPostSubcategory from './blog/computedFields/functions/subcategory';
import buildBlogPostLanguageFlag from './blog/computedFields/functions/language';
import buildLandingPageSlug from './landing-pages/computedFields/functions/slug';
import buildLandingPageUrl from './landing-pages/computedFields/functions/url';
import buildBlogPostCategory from './blog/computedFields/functions/category';
import buildBlogPostSlug from './blog/computedFields/functions/slug';
import buildBlogPostUrl from './blog/computedFields/functions/url';

export {
  buildLandingPageLanguageFlag,
  buildBlogPostLanguageFlag,
  buildLandingPageCategory,
  buildBlogPostSubcategory,
  buildBlogPostCategory,
  buildLandingPageSlug,
  buildLandingPageUrl,
  buildBlogPostSlug,
  buildBlogPostUrl
};
