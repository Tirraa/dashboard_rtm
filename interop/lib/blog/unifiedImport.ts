import isValidLanguageFlag from '../../../src/lib/portable/i18n/isValidLanguageFlag';
import capitalize from '../../../src/lib/portable/str/capitalize';
import indexOfNthOccurrence from '../../../src/lib/portable/str/indexOfNthOccurrence';
import { BLOG_POSTS_FOLDER } from '../../config/blog/contentlayerConfigTweakers';
import { DEFAULT_LANGUAGE } from '../../config/i18n';
import ROUTES_ROOTS from '../../config/routes';
import InvalidArgumentsError from '../../errors/InvalidArguments';
import getFlattenedPathWithoutRootFolder from './builders/computedFields/getFlattenedPathWithoutRootFolder';

export {
  BLOG_POSTS_FOLDER,
  DEFAULT_LANGUAGE,
  InvalidArgumentsError,
  ROUTES_ROOTS,
  capitalize,
  getFlattenedPathWithoutRootFolder,
  indexOfNthOccurrence,
  isValidLanguageFlag
};
