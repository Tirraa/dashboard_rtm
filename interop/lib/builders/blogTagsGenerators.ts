import type { BlogTag } from '../../config/contentlayer/blog/blogTags';

import { blogTagOptions } from '../../config/contentlayer/blog/blogTags';

type EmptyString = '';
const emptyString: EmptyString = '';

export const generateIndexedBlogTagOptions = (__BLOG_TAG_OPTIONS: readonly BlogTag[] = blogTagOptions): Record<BlogTag, number> =>
  __BLOG_TAG_OPTIONS.reduce(
    (acc, tag, index) => {
      acc[tag] = index;
      return acc;
    },
    {} as Record<string, number>
  ) as Record<BlogTag, number>;

export const generateBlogTagOptionsVocabSchema = (__BLOG_TAG_OPTIONS: readonly BlogTag[] = blogTagOptions): Record<BlogTag, EmptyString> =>
  __BLOG_TAG_OPTIONS.reduce(
    (acc, tag) => {
      acc[tag] = emptyString;
      return acc;
    },
    {} as Record<string, string>
  ) as Record<BlogTag, EmptyString>;
