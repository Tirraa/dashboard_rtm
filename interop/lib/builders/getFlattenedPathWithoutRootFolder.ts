import { InvalidArgumentsError } from './unifiedImport';

/**
 * @throws {InvalidArgumentsError}
 */
function getFlattenedPathWithUnkownRootFolder(flattenedPath: string) {
  const firstSlashIndex = flattenedPath.indexOf('/');
  // eslint-disable-next-line @typescript-eslint/no-magic-numbers
  if (firstSlashIndex === -1 || flattenedPath.length - 1 <= firstSlashIndex) {
    throw new InvalidArgumentsError(
      getFlattenedPathWithUnkownRootFolder.name,
      { flattenedPath },
      // eslint-disable-next-line @typescript-eslint/no-magic-numbers
      (firstSlashIndex === -1 ? "Can't find any '/' in flattenedPath" : "Can't find anything after the first '/' in flattenedPath") +
        ". Maybe you just don't need to use this function?"
    );
  }
  // eslint-disable-next-line @typescript-eslint/no-magic-numbers
  return flattenedPath.substring(firstSlashIndex + 1);
}

/**
 * @throws {InvalidArgumentsError}
 */
function getFlattenedPathWithoutRootFolder(flattenedPath: string, rootFolderNeedle?: string) {
  if (!rootFolderNeedle) return getFlattenedPathWithUnkownRootFolder(flattenedPath);

  const expectedFlattenedPathStartStr = rootFolderNeedle + '/';
  if (!flattenedPath.startsWith(expectedFlattenedPathStartStr)) {
    throw new InvalidArgumentsError(
      getFlattenedPathWithoutRootFolder.name,
      { rootFolderNeedle, flattenedPath },
      `flattenedPath doesn't start with "${expectedFlattenedPathStartStr}". You should change the rootFolderNeedle value. Or maybe you just don't need to use this function?`
    );
  }

  return flattenedPath.substring(expectedFlattenedPathStartStr.length);
}

export default getFlattenedPathWithoutRootFolder;
