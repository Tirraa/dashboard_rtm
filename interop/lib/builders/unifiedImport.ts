import { throwIfForbiddenToUseIndexErrorBlogCtx, throwIfForbiddenToUseIndexErrorLpCtx, getPathWithIndexSuffix } from './helpers';
import { LANDING_PAGES_FOLDER, BLOG_POSTS_FOLDER, PAGES_FOLDER } from '../../config/contentlayer/contentlayerConfigTweakers';
import getPathWithoutExtension from '../../../packages/shared-lib/src/portable/str/getPathWithoutExtension';
import { indexedBlogTagOptions, blogTagOptions } from '../../config/contentlayer/blog/blogTags';
import indexOfNthOccurrence from '../../../src/lib/portable/str/indexOfNthOccurrence';
import isValidLanguageFlag from '../../../src/lib/portable/i18n/isValidLanguageFlag';
import getFlattenedPathWithoutRootFolder from './getFlattenedPathWithoutRootFolder';
import ForbiddenToUseIndexError from '../../errors/ForbiddenToUseIndex';
import capitalize from '../../../src/lib/portable/str/capitalize';
import InvalidArgumentsError from '../../errors/InvalidArguments';
import { INDEX_TOKEN } from '../misc/contentlayerCornerCases';
import InvalidBlogTag from '../../errors/InvalidBlogTag';
import { DEFAULT_LANGUAGE } from '../../config/i18n';
import ROUTES_ROOTS from '../../config/routes';

// https://github.com/contentlayerdev/contentlayer/issues/608
export {
  throwIfForbiddenToUseIndexErrorBlogCtx,
  throwIfForbiddenToUseIndexErrorLpCtx,
  getFlattenedPathWithoutRootFolder,
  ForbiddenToUseIndexError,
  getPathWithoutExtension,
  getPathWithIndexSuffix,
  InvalidArgumentsError,
  indexedBlogTagOptions,
  indexOfNthOccurrence,
  LANDING_PAGES_FOLDER,
  isValidLanguageFlag,
  BLOG_POSTS_FOLDER,
  DEFAULT_LANGUAGE,
  InvalidBlogTag,
  blogTagOptions,
  PAGES_FOLDER,
  ROUTES_ROOTS,
  INDEX_TOKEN,
  capitalize
};
