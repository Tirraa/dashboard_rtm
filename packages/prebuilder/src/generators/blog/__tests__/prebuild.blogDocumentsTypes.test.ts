// eslint-disable-next-line import/no-extraneous-dependencies
import { INDEX_TOKEN } from '##/lib/builders/unifiedImport';
// eslint-disable-next-line import/no-extraneous-dependencies
import { afterAll, describe, expect, it } from 'vitest';

import generateBlogDocumentsTypes from '../blogDocumentsTypes';

// https://github.com/vitest-dev/vitest/discussions/2484
const fs = require('fs/promises');

const __TARGET_FOLDER_ROOT = './packages/prebuilder/src/generators/blog/__tests__/FAKE_CODEGEN';
const __TARGET_FOLDER = __TARGET_FOLDER_ROOT + '/' + 'BLOG_DOCUMENTS_TYPES';

describe('generateBlogDocumentsTypes (formatted)', () => {
  afterAll(async () => {
    await fs.rm(__TARGET_FOLDER, { recursive: true });
  });

  const pretty = true;
  it('should match snapshot', async () => {
    const targetFile = 'FAKE_EMPTY_BLOG_DOCUMENTS_TYPES';
    await generateBlogDocumentsTypes({}, pretty, targetFile, __TARGET_FOLDER);

    const fileContent = await fs.readFile(`${__TARGET_FOLDER}/${targetFile}.ts`, 'utf8');
    expect(fileContent).toMatchSnapshot();
  });

  it('should match snapshot', async () => {
    const targetFile = 'FAKE_BLOG_DOCUMENTS_TYPES';
    await generateBlogDocumentsTypes(
      {
        'fake-category-one': {
          'fake-subcategory-one': {
            DEFAULT_LANGUAGE: ['fake-post-01', 'fake-post-02', 'fake-post-03'],
            en: ['fake-post-01', 'fake-post-03'],
            it: ['fake-post-01', 'fake-post-02'],
            fr: ['fake-post-01']
          },
          'fake-subcategory-three': {
            DEFAULT_LANGUAGE: ['fake-post-01', 'fake-post-02', 'fake-post-03', INDEX_TOKEN],
            es: ['fake-post-01', INDEX_TOKEN],
            zh: ['fake-post-02']
          },
          'fake-subcategory-two': { fr: ['fake-post-02'] }
        },
        'fake-category-three': {
          'fake-subcategory-five': { DEFAULT_LANGUAGE: ['fake-post-01', 'fake-post-02', 'fake-post-03'], en: ['fake-post-03'], it: ['fake-post-03'] },
          'fake-subcategory-four': { DEFAULT_LANGUAGE: ['fake-post-01', 'fake-post-02', 'fake-post-03'], es: ['fake-post-04'] }
        },
        'fake-empty-subcategories-nested-in-category': {
          'fake-subcategory-seven': { DEFAULT_LANGUAGE: ['fake-post-01', 'fake-post-02', 'fake-post-03'], zh: ['fake-post-10'] },
          'fake-empty-subcategory': {}
        },
        'fake-category-two': { 'fake-subcategory-six': { DEFAULT_LANGUAGE: ['fake-post-01', 'fake-post-02', 'fake-post-03'], zh: ['fake-post-10'] } },
        'fake-empty-category': {}
      },
      pretty,
      targetFile,
      __TARGET_FOLDER
    );

    const fileContent = await fs.readFile(`${__TARGET_FOLDER}/${targetFile}.ts`, 'utf8');
    expect(fileContent).toMatchSnapshot();
  });
});

describe('generateBlogDocumentsTypes (ugly)', () => {
  afterAll(async () => {
    await fs.rm(__TARGET_FOLDER, { recursive: true });
  });

  const pretty = false;
  it('should match snapshot', async () => {
    const targetFile = 'FAKE_EMPTY_BLOG_DOCUMENTS_TYPES';
    await generateBlogDocumentsTypes({}, pretty, targetFile, __TARGET_FOLDER);

    const fileContent = await fs.readFile(`${__TARGET_FOLDER}/${targetFile}.ts`, 'utf8');
    expect(fileContent).toMatchSnapshot();
  });

  it('should match snapshot', async () => {
    const targetFile = 'FAKE_BLOG_DOCUMENTS_TYPES';
    await generateBlogDocumentsTypes(
      {
        'fake-category-one': {
          'fake-subcategory-one': {
            DEFAULT_LANGUAGE: ['fake-post-01', 'fake-post-02', 'fake-post-03', INDEX_TOKEN],
            en: ['fake-post-01', 'fake-post-03', INDEX_TOKEN],
            it: ['fake-post-01', 'fake-post-02'],
            fr: ['fake-post-01']
          },
          'fake-subcategory-three': {
            DEFAULT_LANGUAGE: ['fake-post-01', 'fake-post-02', 'fake-post-03'],
            es: ['fake-post-01'],
            zh: ['fake-post-02']
          },
          'fake-subcategory-two': { fr: ['fake-post-02'] }
        },
        'fake-category-three': {
          'fake-subcategory-five': {
            DEFAULT_LANGUAGE: ['fake-post-01', 'fake-post-02', 'fake-post-03', INDEX_TOKEN],
            en: ['fake-post-03'],
            it: ['fake-post-03']
          },
          'fake-subcategory-four': {
            DEFAULT_LANGUAGE: ['fake-post-01', 'fake-post-02', 'fake-post-03', INDEX_TOKEN],
            es: ['fake-post-04', INDEX_TOKEN]
          }
        },
        'fake-empty-subcategories-nested-in-category': {
          'fake-subcategory-seven': {
            DEFAULT_LANGUAGE: ['fake-post-01', 'fake-post-02', 'fake-post-03', INDEX_TOKEN],
            zh: ['fake-post-10', INDEX_TOKEN]
          },
          'fake-empty-subcategory': {}
        },
        'fake-category-two': {
          'fake-subcategory-six': { DEFAULT_LANGUAGE: ['fake-post-01', 'fake-post-02', 'fake-post-03', INDEX_TOKEN], zh: ['fake-post-10'] }
        },
        'fake-empty-category': {}
      },
      pretty,
      targetFile,
      __TARGET_FOLDER
    );

    const fileContent = await fs.readFile(`${__TARGET_FOLDER}/${targetFile}.ts`, 'utf8');
    expect(fileContent).toMatchSnapshot();
  });
});
