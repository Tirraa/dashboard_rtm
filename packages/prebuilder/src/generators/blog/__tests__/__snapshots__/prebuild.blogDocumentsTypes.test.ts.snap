// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generateBlogDocumentsTypes (formatted) > should match snapshot 1`] = `
"// AUTOGENERATED: Don't edit this file!
const FAKE_EMPTY_BLOG_DOCUMENTS_TYPES = {} as const;
export default FAKE_EMPTY_BLOG_DOCUMENTS_TYPES;
"
`;

exports[`generateBlogDocumentsTypes (formatted) > should match snapshot 2`] = `
"// AUTOGENERATED: Don't edit this file!
const FAKE_BLOG_DOCUMENTS_TYPES = {
  "FakeCategoryOnePost": {
    "categoryFolder": "fake-category-one"
  },
  "FakeCategoryThreePost": {
    "categoryFolder": "fake-category-three"
  },
  "FakeEmptySubcategoriesNestedInCategoryPost": {
    "categoryFolder": "fake-empty-subcategories-nested-in-category"
  },
  "FakeCategoryTwoPost": {
    "categoryFolder": "fake-category-two"
  },
  "FakeEmptyCategoryPost": {
    "categoryFolder": "fake-empty-category"
  }
} as const;
export default FAKE_BLOG_DOCUMENTS_TYPES;
"
`;

exports[`generateBlogDocumentsTypes (ugly) > should match snapshot 1`] = `
"// AUTOGENERATED: Don't edit this file!
const FAKE_EMPTY_BLOG_DOCUMENTS_TYPES = {} as const;
export default FAKE_EMPTY_BLOG_DOCUMENTS_TYPES;"
`;

exports[`generateBlogDocumentsTypes (ugly) > should match snapshot 2`] = `
"// AUTOGENERATED: Don't edit this file!
const FAKE_BLOG_DOCUMENTS_TYPES = {"FakeCategoryOnePost":{"categoryFolder":"fake-category-one"},"FakeCategoryThreePost":{"categoryFolder":"fake-category-three"},"FakeEmptySubcategoriesNestedInCategoryPost":{"categoryFolder":"fake-empty-subcategories-nested-in-category"},"FakeCategoryTwoPost":{"categoryFolder":"fake-category-two"},"FakeEmptyCategoryPost":{"categoryFolder":"fake-empty-category"}} as const;
export default FAKE_BLOG_DOCUMENTS_TYPES;"
`;
